### DISCUSSION TITLE
Document how Radicale can be sandboxed

### DISCUSSION THREAD

**Discussion by hdatma** [Kozea/Radicale#1423] [General]
Hello, 
#1: Python is a knife without safe handle: it makes it easy to cut yourself. A single line of Python at the unix terminal is good enough to turn any folder into a web page. How do you sandbox Radicale?
#2: Python uses 100MB of disk space. PyPy uses 99MB of disk space. How do you compile Radicale into a stand-alone installation?

**Comment by Unrud** [comment]
1. Radicale has no sandboxing built-in. Just use common practices.
   - Run it as an unprivileged user, that only has write access to the collection.
   - Make also sure that it can't read sensitive files (like home directories if it's running on your personal computer).
   - Be aware that the built-in webserver doesn't enforce any limits on connections, request sizes etc. and thus is prone to DOS attacks. You maybe want to set limits for CPU and RAM usage to protect the rest of the system from slowing down or even better use are real webserver.
2. CPython uses 25 MB on Debian. You could further reduce this by cherry-picking modules from the standard library.

**Comment by hdatma** [comment]
1. When using python to serve an internet port, strong sandboxing/chrooting should be in the project. 
2. Python/CPython on OSX takes exactly 101.8 MB. Did anybody run Radicale under cython.org safely?

**Comment by kousu** [comment]
I see this as a documentation issue: document how to sandbox radicale under Linux, OS X, and FreeBSD and call it a day. Process privilege separation is a thing OSes are good at! Don't reinvent the wheel, etc.

**Comment by liZe** [comment]
I keep point 1 as a documentation issue.

**Comment by stephane-martin** [comment]
Systemd is your friend: http://0pointer.net/public/systemd-nluug-2014.pdf

I'll post an example later :)

**Comment by Unrud** [comment]
> Systemd is your friend:

@hdatma said that he is using OSX. Having configuration examples for different init systems would be nice though..

**Comment by untitaker** [comment]
I don't see why Radicale is responsible for documenting this.

**Comment by hdatma** [comment]
Opened in 2016, this issue is still unresolved in 2024. It is now evident that Radicale is a LAN-only service, with lots of risk mitigations that must be applied but are still to be identified and resolved by the project. Delegating risk mitigation to the user, to the point that the documentation itself is missing a section on this problem, is a no go for me and a solid barrier to adoption.

**Comment by untitaker** [comment]
radicale clearly documents which ports it listens on. examples for various OSes on how to block ports are nice to have (mastodon has them) but absolutely not a requirement, and extremely rare in webservice docs in general. if you rely on pre-made iptables rules to secure your system adequately you should not expose it to the internet in the first place. LAN is less risky.

**Comment by pbiering** [comment]
Nowadays, `radicale` is usually started by `systemd` and the example for the unit file configures already a bunch of security options:

https://github.com/Kozea/Radicale/blob/master/DOCUMENTATION.md
  (search for systemd)

Or use `radicale` in a container.

@hdatma : if this is not enough level of security (nobody else claimed about so far), then please provide documentation how a higher security level can be reached using `radicale` or choose a different software which covers all your security needs.

**Comment by ddubnyk** [comment]
I will provide the options I am using with systemd to effectively sandbox radicale. I am having no issues with it running and functioning as intended. My contacts and calendars all sync correctly.

Of note, to use 

PrivateNetwork=yes

I use a systemd proxy to the private network port. 

My unit file is this:

[Service]
User=radicale
Group=radicale
ReadWritePaths=/var/lib/radicale/
ExecStart=/var/lib/radicale/server/bin/python3 -m radicale
TimeoutStopSec=25s
ProtectSystem=strict
Restart=on-failure
MemoryHigh=256M
SystemCallArchitectures=native
LockPersonality=yes
MemoryDenyWriteExecute=yes
PrivateDevices=yes
ProtectControlGroups=yes
ProtectHome=yes
ProtectHostname=yes
ProtectKernelLogs=yes
ProtectKernelModules=yes
ProtectKernelTunables=yes
RestrictNamespaces=yes
RestrictRealtime=yes
SystemCallFilter=@system-service
PrivateTmp=yes
NoNewPrivileges=true
UMask=0027


I have an override file with these options as well:

[Service]
SystemCallFilter=~@resources

SystemCallFilter=~@obsolete

SystemCallFilter=~@mount

SystemCallFilter=~@debug

SystemCallFilter=~@cpu-emulation

SystemCallFilter=~@privileged

CapabilityBoundingSet=

RemoveIPC=yes

AmbientCapabilities=

ProtectClock=yes

RestrictSUIDSGID=yes

ProtectProc=invisible

ProcSubset=pid

PrivateUsers=yes

PrivateNetwork=yes

IPAddressAllow=localhost

IPAddressDeny=any

RestrictAddressFamilies=~AF_PACKET
RestrictAddressFamilies=AF_UNIX
RestrictAddressFamilies=~AF_NETLINK
RestrictAddressFamilies=~AF_BLUETOOTH
RestrictAddressFamilies=~AF_CAN
RestrictAddressFamilies=~AF_DECnet
RestrictAddressFamilies=~AF_KEY
RestrictAddressFamilies=~AF_RDS
RestrictAddressFamilies=~AF_SNA
RestrictAddressFamilies=~AF_XDP
RestrictAddressFamilies=AF_INET

In the end this is the result of "systemd-analyze security radicale.service"

Overall exposure level for radicale.service: 0.5 SAFE 😀

**Comment by pbiering** [comment]
By https://github.com/Kozea/Radicale/pull/2198 a service file is now shipped with upstream in

https://github.com/Kozea/Radicale/tree/master/contrib/systemd

I will check on my servers whether part of your additional restrictions can be applied as well.
